File: TP/JAVA/PalindromoRecursivo.java
/*Crie um método recursivo que recebe uma string como parâmetro e retorna true se essa é um ``Palíndromo''. Na saída padrão, para cada linha de entrada, escreva uma linha de saída com SIM/NÃO indicando se a linha é um palíndromo. Destaca-se que uma linha de entrada pode ter caracteres não letras.
*/

import java.util.*;


public class PalindromoRecursivo{
  
  public static void palindromo(String frase, int inicio, int fim){
    if(inicio >= fim){ 
        System.out.println("SIM");
        return;
    }
    if(frase.charAt(inicio) != frase.charAt(fim)){ //compara o char. se for diferente a palavra nao eh palindromo
      System.out.println("NAO");
      return;
    }
    palindromo(frase, inicio +1, fim -1); //chama novamente o metodo, andando para a prox letra
  }

  public static void main(String[] args){
    Scanner sc = new Scanner(System.in);
    String frase;
    int tam;
    
    while(sc.hasNext()){
      frase = sc.nextLine();
      
       if ((frase.length() == 3) && (frase.charAt(0) == 'F') && (frase.charAt(1) == 'I') && (frase.charAt(2) == 'M')){ //acaba programa se receber a palavra FIM
        sc.close();
        return;
      }
      
      tam = frase.length();
      palindromo(frase, 0, tam - 1);   //chama o metodo recursivo 
    }
  }
}

File: TP/tp1/C/PalindromoRecursivo.c
/* Crie um método recursiva que recebe uma string como parâmetro e retorna true se essa é um ``Palíndromo''. Na saída padrão, para cada linha de entrada, escreva uma linha de saída com SIM/NÃO indicando se a linha é um palíndromo. Destaca-se que uma linha de entrada pode ter caracteres não letras.
*/

#include <stdio.h>
#include <string.h>


int palindromo(int result, int inicio, int final, char* string){
  if((inicio >= final) || (result == 0)){
    return result; //retorna result
  }
  if(string[inicio] != string[final]){ //compara os char da String
    result = 0;
  }
  palindromo(result,inicio+1,final-1,string); //chama a funcao novamente, andando INICIO para a direita e FINAL para a esquerda
}

int main(){
  int tam, result;
  char string[500];
  while(scanf("%[^\n]",string) != EOF){ //funciona enquanto tiver palavras no arquivo. %[^\n] serve para nao ter problemas com \n
    getchar();
    
    if(strcmp(string, "FIM") == 0){ //funciona enquanto tiver palavras no arquivo
      return 0;
    }
    tam = strlen(string); //define tam como o tamanho do array
    result = palindromo(1,0,tam - 1, string); //chama funcao recursiva
    if(result == 1){ //se result foi retornado 1 a palavra eh palindromo
      printf("SIM\n");
    }else{
      printf("NAO\n");
    }
  }
}
